### DISCUSSION TITLE
ddl-auto의 옵션은 각각 어떤 동작을 하고 어떤 상황에서 사용해야 할까요?

### DISCUSSION THREAD

**Discussion by Kimprodp** [2024-woowacourse-study/level-interview#161] [Spring Data]
.

**Comment by alstn113** [comment]
`ddl-auto` 옵션은 스프링 부트 애플리케이션에서 Hibernate와 같은 JPA 구현체를 사용할 때 데이터베이스 스키마 관리를 제어하는 설정입니다. 이 옵션은 `application.properties` 또는 `application.yml` 파일에서 설정할 수 있으며, 다양한 값에 따라 데이터베이스 스키마에 대해 다른 동작을 수행합니다. 아래는 각 값과 그 동작 및 사용 상황에 대한 설명입니다:

### 1. `none`
- **동작**: Hibernate가 데이터베이스 스키마와 관련된 어떠한 작업도 수행하지 않습니다.
- **사용 상황**: 프로덕션 환경에서 주로 사용됩니다. 데이터베이스 스키마를 수동으로 관리하고 싶을 때 유용합니다.

### 2. `validate`
- **동작**: 애플리케이션이 시작될 때 엔티티 매핑이 데이터베이스 스키마와 일치하는지 검증합니다. 스키마 변경은 수행하지 않습니다.
- **사용 상황**: 프로덕션 환경에서 엔티티와 데이터베이스 스키마가 일치하는지 확인하고 싶을 때 사용합니다. 데이터베이스 스키마가 변경되지 않도록 보장해야 할 때 유용합니다.

### 3. `update`
- **동작**: 엔티티 매핑과 데이터베이스 스키마를 비교하여 필요한 경우 스키마를 업데이트합니다. 기존 데이터는 유지되지만, 새로운 엔티티나 변경된 엔티티 필드는 스키마에 반영됩니다.
- **사용 상황**: 개발 단계에서 편리하게 사용됩니다. 애플리케이션을 개발하면서 엔티티에 변경이 발생할 때 자동으로 스키마를 업데이트하고 싶을 때 유용합니다. 그러나 프로덕션 환경에서는 예기치 않은 스키마 변경을 방지하기 위해 주의가 필요합니다.

### 4. `create`
- **동작**: 애플리케이션이 시작될 때 기존 스키마를 삭제하고 새로 생성합니다. 데이터가 모두 삭제되며, 엔티티 매핑을 기반으로 새로운 스키마가 생성됩니다.
- **사용 상황**: 초기 개발 단계나 테스트 환경에서 주로 사용됩니다. 개발 초기에 빈 데이터베이스 스키마를 반복적으로 생성해야 할 때 유용합니다. 주의: 기존 데이터가 모두 삭제되므로 프로덕션 환경에서는 사용하지 않습니다.

### 5. `create-drop`
- **동작**: `create`와 유사하지만, 애플리케이션이 종료될 때 스키마를 삭제합니다.
- **사용 상황**: 테스트 환경에서 일시적인 데이터베이스 스키마가 필요한 경우 유용합니다. 매 테스트 실행 시마다 깨끗한 데이터베이스 상태를 유지하고자 할 때 사용됩니다. 프로덕션 환경에서는 사용하지 않습니다.

### 요약
- **개발 단계**: `update`, `create`, `create-drop` 옵션을 사용하여 개발 편의성을 높일 수 있습니다.
- **테스트 단계**: `create-drop` 옵션을 사용하여 매 테스트 실행 시마다 깨끗한 데이터베이스 상태를 유지할 수 있습니다.
- **프로덕션 단계**: `none` 또는 `validate` 옵션을 사용하여 예기치 않은 스키마 변경을 방지하고, 스키마가 애플리케이션의 엔티티 매핑과 일치하는지 검증할 수 있습니다.

스키마 변경이 필요할 때는 적절한 데이터베이스 마이그레이션 도구(Flyway, Liquibase 등)를 사용하여 제어된 방식으로 스키마를 관리하는 것이 좋습니다.

**Comment by 3Juhwan** [comment]
ddl-auto는 스프링 부트가 데이터베이스를 초기화하는 방법 중 hibernate를 사용하는 방법입니다. 엔티티만 등록하면 스키마 자동 생성 및 관리를 제어할 수 있습니다. 

ddl-auto는 다음 5가지 옵션을 갖고 있습니다. 

**none(default)**
- Hibernate가 데이터베이스 스키마와 관련된 어떠한 작업도 수행하지 않습니다.
- embedded 데이터베이스가 아닌 경우에 none이 기본 옵션입니다. 

**validate**
- DDL을 작성하여 테이블을 생성하거나 수정하지 않고, 엔티티 클래스와 테이블이 정상적으로 매핑되는지만 검사합니다. 
- 테이블이 아예 존재하지 않거나, 테이블에 엔티티의 필드에 매핑되는 컬럼이 존재하지 않으면 예외를 발생시키면서 애플리케이션을 종료합니다. 

**update**
- Hibernate가 매핑된 엔티티에 따라 데이터베이스 스키마를 변경합니다. 
- 새로운 테이블 또는 칼럼을 추가하지만 기존의 테이블이나 칼럼을 삭제하거나 변경하지는 않습니다.
- 기존에 존재하는 컬럼의 속성(nullable, 크기, 데이터 타입 등)은 건드리지 않습니다. 가령, 어떠한 엔티티 클래스의 String 필드를 Int로 변경하더라도 해당 엔티티에 매핑되는 테이블의 해당 컬럼은 숫자타입으로 바뀌지 않고 문자열 타입으로 유지합니다. 

**create**
- Hibernate가 애플리케이션을 시작할 때 기존의 데이터베이스 스키마를 삭제하고 새로운 스키마를 생성합니다.

**create-drop**
- create와 유사하지만, 애플리케이션이 종료될 때 데이터베이스 스키마를 삭제합니다.
- embedded 데이터베이스의 경우에 create-drop이 기본 옵션입니다. 

</br>

환경에 따라 권장하는 옵션은 다음과 같습니다. 

**개발 환경**
- create 또는 update
- 개발 중에는 엔티티가 자주 변경될 수 있으므로 create 또는 update를 사용하여 스키마를 동기화합니다.

**테스트 환경**
- update 또는 validate

**운영 환경**
- validate 또는 none
- 운영 환경에서는 데이터 손실을 방지하기 위해 validate로 매핑을 검증하거나, 데이터베이스 스키마 변경을 완전히 막기 위해 none을 사용합니다.

</br>

> [!NOTE]  
> https://www.inflearn.com/course/ORM-JPA-Basic/dashboard
> https://docs.spring.io/spring-boot/how-to/data-initialization.html#howto.data-initialization.using-hibernate
